### ISSUE TITLE
Unable to use done property of IteratorResult to narrow types in typescript 3.6.2

### ISSUE THREAD

**Issue by brainkim** [microsoft/TypeScript#33241]
**TypeScript Version:**  3.6.2

**Search Terms:**
async iterator IteratorResult iterable next return throw type narrowing

**Code**
file: iterator-result.ts
```ts
declare const result: IteratorResult<string, number>;
if (!result.done) {
  let value: string = result.value;
  console.log(value);
}
```

**Expected behavior:**
The code compiles. Checking the negation of `result.done` narrows the type of `result.value` to `string`.

**Actual behavior:**
```
iterator-result.ts(3,7): error TS2322: Type 'string | number' is not assignable to type 'string'.
  Type 'number' is not assignable to type 'string'.
```

I thought the whole point of strictly typed iterators was that we could use the `done` to narrow the type of `value`, but it seems like a late-game change (making  `IteratorYieldResult.done` optional) prevents us from doing this.

I’m fine with having `result.done` be optional for `IteratorYieldResult` but this raises the question why we didn’t also allow `result.value` to be optional in `IteratorReturnResult`, when it is idiomatic javascript to return { done: true } without a value in custom iterators.

**Playground Link:** 

Typescript 3.6.2 is not available in the playground yet.

**Related Issues:** 
https://github.com/microsoft/TypeScript/pull/30790 <- pr which introduced this change.
https://github.com/microsoft/TypeScript/issues/2983
https://github.com/microsoft/TypeScript/issues/10564
https://github.com/Microsoft/TypeScript/issues/11375
https://github.com/microsoft/TypeScript/issues/27059

**Comment by MattiasBuelens** [comment]
According to [this comment on #27059](https://github.com/microsoft/TypeScript/issues/27059#issuecomment-420878991), this bug only happens when `strictNullChecks` is off. If you can, you should enable `strictNullChecks` (or `strict`).

**Comment by brainkim** [comment]
Can confirm, was testing with a plain tsc call which did not enable strict.

**Comment by rbuckton** [comment]
This is currently a limitation of control-flow when `strictNullChecks` is not enabled. The problem is that `null` and `undefined` are assignable to a type of `true`, therefore we cannot remove the `done: true` discriminant from the union in the negative case of `!result.done`. Please note that the following conditions do work:

```ts
if (result.done) {
  let value: number = result.value;
}
if (result.done === false) {
  let value: string = result.value;
}
```
